File: leetcode_ly/java/List/InsertionSortList.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode insertionSortList(ListNode head) {
        if(head == null) return null;
        ListNode node = new ListNode(0);
        while(head != null){
            ListNode pre = node;
            while(pre.next != null && pre.next.val <= head.val){
                pre = pre.next;
            }
            ListNode temp = head.next;
            head.next = pre.next;
            pre.next = head;
            head = temp;
        }
        return node.next;
    }
}

File: leetcode_ly/c++/List/InsertionSortList.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* insertionSortList(ListNode* head) {
        if(!head || !head->next) return head;
        ListNode* node = new ListNode(INT_MIN);
        node->next = NULL;
        while(head){
            ListNode* tail = head->next;
            int index = getIndex(node,head);
            ListNode* front = node;
            for(int i = 0; i < index - 1; i++){
                front = front->next;
            }
            head->next = front->next;
            front->next = head;
            head = tail;
        }
        return node->next;
    }
    
    int getIndex(ListNode* node,ListNode* head){
        int index = 0;
        while(node){
            if(node->val < head->val){
                index++;
            }
            node = node->next;
        }
        return index;
    }
};